scripts/test/generate-image.sh

```diff
@@ -0,0 +1,169 @@
+          --password password --package=openssh-server,dhcpcd5 --package grub-efi-amd64-signed \
+          --package shim-signed --package lsb-release --output="$image" build
+
+    post_process_image "$image"
+
+    echo "Image successfully generated!"
+}
+
+generate_ubuntu() {
+    local image="Ubuntu-Focal-x86-64.img"
+
+    mkosi --root-size=2G --distribution=ubuntu --release=focal --format=gpt_ext4 --bootable --checksum \
+          --password password --package=openssh-server,dhcpcd5 --package grub-efi-amd64-signed \
+          --package shim-signed --package lsb-release --output="$image" build
+
+    post_process_image "$image"
+
+    echo "Image successfully generated!"
+}
+
+post_process_image() {
+    local image="$1"
+
+    mkdir -p tmp-p1
+    mkdir -p tmp-p2
+
+    LO_DEVICE=$(losetup --find --show --partscan "$image")
+    trap cleanup_losetup EXIT
+    mount ${LO_DEVICE}p1 tmp-p1
+    mount ${LO_DEVICE}p2 tmp-p2
+
+    pre_tweaks tmp-p1 tmp-p2
+
+    create_grub_regeneration_service tmp-p1 tmp-p2
+    umount tmp-p1
+    umount tmp-p2
+    regenerate_grub_live "$image"
+    mount ${LO_DEVICE}p1 tmp-p1
+    mount ${LO_DEVICE}p2 tmp-p2
+
+    post_tweaks tmp-p1 tmp-p2
+}
+
+pre_tweaks() {
+    local boot="$1"
+    local root="$2"
+
+    # Fstab is missing for some reason. I'm not exactly sure why systemd-boot
+    # works without this, and GRUB doesn't.
+    cat > "$root/etc/fstab" <<EOF
+/dev/root / auto defaults 0 0
+/dev/sda1 /boot/efi auto defaults 0 0
+EOF
+
+    # Real installers create a /etc/default/grub file with a distributor in
+    # them.
+    if [ ! -e "$root/etc/default/grub" ]; then
+        mkdir -p $root/etc/default
+        echo 'GRUB_DISTRIBUTOR=`lsb_release -i -s 2> /dev/null`' > $root/etc/default/grub
+    fi
+
+    sed -E -i -e 's/^#? *PermitRootLogin .*/PermitRootLogin yes/' $root/etc/ssh/sshd_config
+}
+
+post_tweaks() {
+    local boot="$1"
+    local root="$2"
+
+    # Delete systemd-boot, which isn't normally present in images that were
+    # installed with OS installers, at least not at the time of writing.
+    rm -rf "$boot/EFI/systemd"
+
+    # Also replace bootx64.efi, which is the default bootloader. Mkosi installs
+    # systemd-bootx86.efi, but we want the shim.
+    rm -f "$boot/EFI/BOOT/*"
+    mkdir -p "$boot/EFI/BOOT"
+    cp "$root/usr/lib/shim/shimx64.efi.signed" "$boot/EFI/BOOT/BOOTX64.EFI"
+}
+
+# Unfortunately installing grub scripts is something which is not really
+# possible when offline. This is something which is easier with systemd-boot, so
+# longterm GRUB will probably follow, or systemd-boot will take over. Anyway,
+# let's do it by using a systemd service to perform the job, and then shut down.
+create_grub_regeneration_service() {
+    local boot="$1"
+    local root="$2"
+
+    cat > "$root/etc/systemd/system/mender-regenerate-grub-and-shutdown.service" <<EOF
+[Unit]
+Description=Regenerate grub scripts, disable itself and then shut down.
```

Comment:
Yeah. In retrospect, it may have been better to use the `chroot` approach here as well, but I did this long before I got to the `mender-convert` part of it. It's quite a lot of trial and error to change it now, and it's only for internal use, so I think I'll leave this for now.

Reply:
Yeh, it is fine imo